src/common/column_matrix.h

```diff
@@ -13,310 +13,491 @@
+    return sparse_clmn_ptr_ ? ColumnType::kSparseColumn : ColumnType::kDenseColumn;
+  }
+
+  size_t Size() const {
+    return sparse_clmn_ptr_ ? sparse_clmn_ptr_->Size() : dense_clmn_ptr_->Size();
+  }
+
+  const size_t* GetRowData() const {
+    return sparse_clmn_ptr_ ? sparse_clmn_ptr_->GetRowData() : nullptr;
+  }
+
+  template <typename BinIdxType, typename CastType = Column::BinCmpType>
+  CastType GetBinIdx(size_t rid, size_t* state) const {
+    return sparse_clmn_ptr_ ? sparse_clmn_ptr_->GetBinIdx<BinIdxType, CastType>(rid, state)
+                            : dense_clmn_ptr_->GetBinIdx<BinIdxType, CastType>(rid, state);
+  }
+
+  /*
+   * Additional method for maintaining the backward compatibility 
+   * with the non-optimized partition builder
+   * TODO: Delete after the optimizied partition builder will be implemented
+   */
+  template <typename CastType = Column::BinCmpType>
+  CastType GetBinId(size_t rid, size_t* state) const {
+    return sparse_clmn_ptr_ ? sparse_clmn_ptr_->GetBinId<CastType>(rid, state)
+                            : dense_clmn_ptr_->GetBinId<CastType>(rid, state);
   }
+
+  size_t GetInitialState(const size_t first_row_id) const {
+    return sparse_clmn_ptr_ ? sparse_clmn_ptr_->GetInitialState(first_row_id)
+                            : dense_clmn_ptr_->GetInitialState(first_row_id);
+  }
+
+  size_t GetRowIdx(size_t idx) const {
+    return sparse_clmn_ptr_ ? sparse_clmn_ptr_->GetRowIdx(idx) : 0;
+  }
+
+ private:
+  const SparseColumn * sparse_clmn_ptr_;
+  const DenseColumn* dense_clmn_ptr_;
 };
 
-/**
- * \brief Column major matrix for gradient index. This matrix contains both dense column
- * and sparse column, the type of the column is controlled by sparse threshold. When the
- * number of missing values in a column is below the threshold it's classified as dense
- * column.
- */
+/*! \brief a collection of columns, with support for construction from
+    GHistIndexMatrix. */
 class ColumnMatrix {
  public:
+  using ByteType = uint8_t;
+  using ColumnListType = std::vector<std::shared_ptr<const Column>>;
+  using ColumnViewListType = std::vector<std::shared_ptr<const ColumnView>>;
   // get number of features
   bst_feature_t GetNumFeature() const { return static_cast<bst_feature_t>(type_.size()); }
+  // get index data ptr
+  template <typename Data>
+  const Data* GetIndexData() const {
+    return reinterpret_cast<const Data*>(index_.data());
+  }
 
-  template <typename Batch>
-  void Init(Batch const& batch, float missing, GHistIndexMatrix const& gmat,
-            double sparse_threshold, int32_t n_threads) {
-    auto const nfeature = static_cast<bst_feature_t>(gmat.cut.Ptrs().size() - 1);
-    const size_t nrow = gmat.row_ptr.size() - 1;
+  // get index data ptr
+  const ByteType* GetIndexData() const {
+    return index_.data();
+  }
+
+  // construct column matrix from GHistIndexMatrix
+  inline void Init(SparsePage const& page, const GHistIndexMatrix& gmat, double sparse_threshold,
```

Comment:
The `Batch` is for preparing the column matrix to use adapters https://github.com/dmlc/xgboost/blob/master/src/data/adapter.h . We want to avoid SparsePage as it might consume 3 times more memory than the original data.